### PR TITLE
Comment on solution

### PR DESCRIPTION
**Pull request by itraviv**

### CODE REVIEW

File: lmsweb/views.py [hunk 1]
```diff
@@ -147,31 +154,79 @@ def exercises_page():
...
+        return fail(404, f"No such solution {solution_id}")
+
+    solver_id = solution.solver.id
+    if solver_id != session_id:
+        return fail(401, "You aren't allowed to watch this page.")
+
+    if act == 'fetch':
+        return jsonify(Comment.by_solution(solution_id))
+
+    if session['role'] in HIGH_ROLES:
```
Comment (yammesicka) [inline lmsweb/views.py]: not in

File: lmsweb/views.py [hunk 2]
```diff
@@ -147,31 +154,79 @@ def exercises_page():
...
-            comment_id = int(request.args.get('commentId'))
-            # TODO: Handle if not found
-            Comment.get(
-                Comment.comment == comment_id,
-                Comment.solution == solution_id,
-            ).delete_instance()
-            return jsonify({"success": "true"})
+    act = request.args.get('act')
+
+    solution_id = int(request.form.get('solutionId', 0))
```
Comment (yammesicka) [inline lmsweb/views.py]: Will work on POST but not on GET

File: lmsweb/views.py [hunk 3]
```diff
@@ -147,31 +154,79 @@ def exercises_page():
     )
 
 
+def _create_comment(session_id: int, solution: Solution):
+    user = User.get_or_none(User.id == session_id)
```
Comment (yammesicka) [inline lmsweb/views.py]: Receive the argument with the name `user_id`

File: lmsweb/views.py [hunk 4]
```diff
@@ -147,31 +154,79 @@ def exercises_page():
...
+    if user is None:
+        # should never happen, we checked session_id == solver_id
+        return fail(404, 'No such user')
+
+    kind = request.form.get('kind', '')
```
Comment (yammesicka) [inline lmsweb/views.py]: Pass as a parameter

File: lmsweb/views.py [hunk 5]
```diff
@@ -147,31 +154,79 @@ def exercises_page():
...
+    try:
+        line_number = int(request.form.get('line', 0))
+    except ValueError:
+        line_number = 0
+    if line_number <= 0:
+        return fail(422, f"Invalid line number: {line_number}")
+
+    if kind == 'id':
+        comment_id = int(request.form.get('comment', 0))
+
```
Comment (yammesicka) [inline lmsweb/views.py]: Linebreak is not needed

File: lmsweb/views.py [hunk 6]
```diff
@@ -147,31 +154,79 @@ def exercises_page():
...
+        return fail(400, "Invalid kind")
+
+    Comment.create(
+        commenter=user,
+        timestamp=datetime.now(),
+        line_number=line_number,
+        comment=comment_id,
+        solution=solution
+    )
+    return jsonify({"success": "true"})
```
Comment (yammesicka) [inline lmsweb/views.py]: Return the fields `id` and `text` with the right values. The frontend depends on it to update the view.

File: lmsweb/views.py [hunk 7]
```diff
@@ -147,31 +154,79 @@ def exercises_page():
...
+    if user is None:
+        # should never happen, we checked session_id == solver_id
+        return fail(404, 'No such user')
+
+    kind = request.form.get('kind', '')
+    if (not kind) or (kind not in ('id', 'text')):
+        return fail(400, "Invalid kind")
+
+    try:
+        line_number = int(request.form.get('line', 0))
```
Comment (yammesicka) [inline lmsweb/views.py]: Pass as a parameter (this function should not know about the form or how the values are fetched)

File: lmsweb/views.py [hunk 8]
```diff
@@ -147,31 +154,79 @@ def exercises_page():
...
+
+    if act == 'fetch':
+        return jsonify(Comment.by_solution(solution_id))
+
+    if session['role'] in HIGH_ROLES:
+        return fail(401, "You must be an admin to view this page.")
+
+    if act == 'delete':
+        comment_id = int(request.args.get('commentId'))
+        comment_ = Comment.get_or_none(Comment.id == comment_id)
```
Comment (yammesicka) [inline lmsweb/views.py]: just `comment` will do actually

Comment (itraviv) [reply lmsweb/views.py]: shadowing the `comment` function name. that why i used _

File: lmsweb/views.py (hunk 3 above)
Comment (itraviv) [reply lmsweb/views.py]: but this actually the row from the db. not the id

File: lmsweb/views.py [hunk 9]
```diff
@@ -147,31 +153,100 @@ def exercises_page():
...
+    else:  # it's a GET
+        solution_id = int(request.args.get('solutionId', 0))
+    session_id = int(session['id'])
+
+    solution = Solution.get_or_none(Solution.id == solution_id)
+    if solution is None:
+        return fail(404, f"No such solution {solution_id}")
+
+    solver_id = solution.solver.id
+    if solver_id != session_id:
```
Comment (yammesicka) [inline lmsweb/views.py]: Allow high roles to view the page

**Pull request merged**